File: leap.js
//NOTE(adam): make a function that prints the next 20 leap years
const getNextLeap = (year) => {
  const next = year + (4 - parseInt(year % 4))
  return (next % 100 === 0 && next % 400 !== 0) ? next + 4 : next
}

//NOTE(adam): generator for next 'count' leap years, not considering current
function*  makeLeapGen(startYear, count) {
  //NOTE(adam): optional count argument default to 20
  count = count || 20

  let next = startYear  //NOTE(adam): cleanest implementation allows mutation
  let curCount = 0      //NOTE(adam): allow mutation for counter variable
  while(curCount++ < count) {
    //NOTE(adam): skip year if is multiple of 100
    yield next = getNextLeap(next)
  }
}

//NOTE(adam): use generator to return a list
const makeLeapList = (startYear, count) => [...makeLeapGen(startYear, count)]


console.log('[...makeLeapGen(1899)]', [...makeLeapGen(1899)])
console.log('[...makeLeapGen(1999)]', [...makeLeapGen(1999)])
console.log('[...makeLeapGen(2020, 5)]', [...makeLeapGen(2020, 5)])
console.log('makeLeapList(1940)', makeLeapList(1940))

File: leap.clj
;NOTE(adam): make a function that prints the next 20 leap years

;NOTE(adam): assume year is multiple of 4
;            false if  multiple of 100 unless multiple of 400
(defn isLeap [year]
  (not (and
    (= (mod year 100) 0)
    (not (= (mod year 400) 0)))))

;NOTE(adam): get leap year following argument
(defn getNextLeap [year]
  (def nextLeap
    (+ year (- 4 (mod year 4))))
  (if (not (isLeap nextLeap))
    (+ nextLeap 4)
    nextLeap))

;NOTE(adam): get list of leap years
(defn leaps
  ([startYear] (take 20 (iterate getNextLeap startYear)))
  ([startYear yearCount] (take yearCount (iterate getNextLeap startYear))))

(println (leaps 1899))
(println (leaps 1999))
(println (leaps 2020 5))
(println (leaps 1940))
